### PR TITLE
Configure Token provider from the corresponding provider

### PR DESCRIPTION
**Pull request by MikeAlhayek**
Enhanced functionality has been introduced, empowering developers to manage the expiration time of various tokens, including password-reset, email-confirmation, change-email, and two-factor authentication, delivered via the email service. Below, you'll find a list of configurable options along with their default values:
 
 | Class Name | Default Expiration Value |
 | ---------- | ------------------------ |
 | `ChangeEmailTokenProviderOptions` | The token is valid by default for **15 minutes**. |
 | `EmailConfirmationTokenProviderOptions` | The token is valid by default for **48 hours**. |
 | `PasswordResetTokenProviderOptions` | The token is valid by default for **15 minutes**. |

 You many change the default values of these options by using the `services.Configure<>` method. For instance, to change the `EmailConfirmationTokenProviderOptions` you can add the following code to your project

 ```csharp
 services.Configure<EmailConfirmationTokenProviderOptions>(options => options.TokenLifespan = TimeSpan.FromDays(7));
 ```

### Security Upgrade
This PR also enhance the security of the app by increasing the length of the reset-password token. Currently the rest-password token is only 6 characters in length. With this update, it is a long token and can't be guessed as easy as 6 characters.

### CODE REVIEW

File: src/OrchardCore/OrchardCore.Users.Core/Extensions/UsersServiceCollectionExtensions.cs
Comment (lampersky) [inline src/OrchardCore/OrchardCore.Users.Core/Extensions/UsersServiceCollectionExtensions.cs:71]: please read my comment [here](https://github.com/OrchardCMS/OrchardCore/discussions/15585#discussioncomment-8956352)

**Comment by lampersky** [comment]
this PR looks ok to me, but it is not solving the issue mentioned [here](https://github.com/OrchardCMS/OrchardCore/discussions/15585)

File: src/OrchardCore/OrchardCore.Users.Core/Extensions/UsersServiceCollectionExtensions.cs
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Users.Core/Extensions/UsersServiceCollectionExtensions.cs:71]: @lampersky yes this change will not fix that. The intention was not to fix that. just clean up to when the these providers were registered. I'll reply to your comment in that thread.

File: src/OrchardCore.Modules/OrchardCore.Users/Startup.cs
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Users/Startup.cs:313]:
```suggestion
            services.TryAddDefaultEmailTokenProvider()
                .Configure<IdentityOptions>(options => options.Tokens.ChangeEmailTokenProvider = TokenOptions.DefaultEmailProvider);
```

File: src/OrchardCore.Modules/OrchardCore.Users/Startup.cs
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Users/Startup.cs:374]:
```suggestion
            services.TryAddDefaultEmailTokenProvider()
                .Configure<IdentityOptions>(options => options.Tokens.EmailConfirmationTokenProvider = TokenOptions.DefaultEmailProvider);
```

**Review by hishamco — CHANGES_REQUESTED** [review]

**Comment by hishamco** [comment]
@MikeAlhayek could you please address this issue https://github.com/OrchardCMS/OrchardCore/discussions/15585 please

**Review by hishamco — APPROVED** [review]

**Comment by hishamco** [comment]
LGTM if this will not address the @lampersky issue, it would be nice to address the issue in this PR while you're working on this

**Comment by MikeAlhayek** [comment]
@lampersky @hishamco as a bonus, I implemented a fix for #15585

@sebastienros  @Piedone let me know your thoughts on these default values for OrchardCore. I think these as okay values to have by default. But, would like to know if you have a better fix default values that you like to change to.

New options were added for the email token-providers to give you control on configuring the expiration time on each token sent via email. Here is a list of the available options that you can configure:

 | Class Name | Default Expiration Value |
 | ----------- | ----------- |
 | `PasswordResetTokenProviderOptions` | The token is valid by default for **15 minutes**. |
 | `EmailConfirmationTokenProviderOptions` | The token is valid by default for **48 hours**. |
 | `ChangeEmailTokenProviderOptions` | The token is valid by default for **30 minutes**. |

**Comment by Piedone** [comment]
I'd use the same 15 minutes for `ChangeEmailTokenProviderOptions` as for `PasswordResetTokenProviderOptions`. These open up access to the user account, so they should be as short as possible, while allowing reasonably enough time for an e-mail to arrive and the user to react. 15 minutes is enough for this.

`EmailConfirmationTokenProviderOptions` looks good.

**Comment by MikeAlhayek** [comment]
> I'd use the same 15 minutes for ChangeEmailTokenProviderOptions as for PasswordResetTokenProviderOptions

Not a bad idea. Changed

File: src/docs/releases/1.9.0.md
Comment (hishamco) [inline src/docs/releases/1.9.0.md:239]:
```suggestion
 You may change the default values of these options by using the `services.Configure<>` method. For instance, to change the `EmailConfirmationTokenProviderOptions` you can add the following code to your project
```

File: src/docs/releases/1.9.0.md
Comment (hishamco) [inline src/docs/releases/1.9.0.md:230]:
```suggestion
Enhanced functionality has been introduced, empowering developers to manage the expiration time of various tokens, including password-reset, email-confirmation, change-email, and two-factor authentication, delivered via the email service. Below, you'll find a list of configurable options along with their default values:
```

File: src/OrchardCore/OrchardCore.Users.Core/Services/TwoFactorEmailTokenProvider.cs
Comment (hishamco) [inline src/OrchardCore/OrchardCore.Users.Core/Services/TwoFactorEmailTokenProvider.cs:193]:
```suggestion
        if (var modifierBytes = modifier is not null)
        {
            _encoding.GetBytes(modifier);
        }
```

Comment (hishamco) [reply src/OrchardCore/OrchardCore.Users.Core/Services/TwoFactorEmailTokenProvider.cs:193]: No need for `else` because the `modifier` is `null` by default

File: src/OrchardCore/OrchardCore.Users.Core/Services/TwoFactorEmailTokenProvider.cs
Comment (hishamco) [inline src/OrchardCore/OrchardCore.Users.Core/Services/TwoFactorEmailTokenProvider.cs:143]: Why the assertion is needed?

File: src/OrchardCore/OrchardCore.Users.Core/Services/TwoFactorEmailTokenProvider.cs
Comment (hishamco) [inline src/OrchardCore/OrchardCore.Users.Core/Services/TwoFactorEmailTokenProvider.cs:89]:
```suggestion
    private static async Task<string> GetUserModifierAsync(string purpose, UserManager<IUser> manager, IUser user)
    {
        ArgumentNullException.ThrowIfNull(user);
        var userId = await manager.GetUserIdAsync(user);

        return $"Totp:{purpose}:{userId}";
    }

    private string GetStringFormat()
    {
        // Number of 0's is length of the generated pin.
        _format ??= _options.TokenLength switch
        {
            TwoFactorEmailTokenLength.Two => "D2",
            TwoFactorEmailTokenLength.Three => "D3",
            TwoFactorEmailTokenLength.Four => "D4",
            TwoFactorEmailTokenLength.Five => "D5",
            TwoFactorEmailTokenLength.Six => "D6",
            TwoFactorEmailTokenLength.Seven => "D7",
            TwoFactorEmailTokenLength.Eight or TwoFactorEmailTokenLength.Default => "D8",
            _ => throw new NotSupportedException("Unsupported token length.")
        };

        return _format;
    }
```

File: src/OrchardCore/OrchardCore.Users.Core/Services/TwoFactorEmailTokenProvider.cs
Comment (hishamco) [inline src/OrchardCore/OrchardCore.Users.Core/Services/TwoFactorEmailTokenProvider.cs:17]:
```suggestion
    private readonly TwoFactorEmailTokenProviderOptions _options;
    
    private string _format;
```

**Review by hishamco — CHANGES_REQUESTED** [review]

File: src/OrchardCore/OrchardCore.Users.Core/Services/TwoFactorEmailTokenProvider.cs
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Users.Core/Services/TwoFactorEmailTokenProvider.cs:143]: This code is mainly influenced by the references class. So it's just copied for the most part

File: src/OrchardCore/OrchardCore.Users.Core/Services/TwoFactorEmailTokenProvider.cs
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Users.Core/Services/TwoFactorEmailTokenProvider.cs:89]: What is the change here?

File: src/OrchardCore/OrchardCore.Users.Core/Services/TwoFactorEmailTokenProvider.cs
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Users.Core/Services/TwoFactorEmailTokenProvider.cs]: I thought that :)

**Review by hishamco — APPROVED** [review]

**Comment by hishamco** [comment]
@lampersky your review please before merge

**Comment by lampersky** [comment]
is it finished? I see new commits each day 😜

**Comment by MikeAlhayek** [comment]
not ready yet. I am trying to add some tests for the Rfc6238Authentication service. I am not comfortable with the logic.

**Comment by hishamco** [comment]
Do you need any help?

**Comment by MikeAlhayek** [comment]
@lampersky okay. It's ready now. Feel free to review the code.

File: src/OrchardCore/OrchardCore.Users.Core/Services/Rfc6238AuthenticationService.cs
Comment (lampersky) [inline src/OrchardCore/OrchardCore.Users.Core/Services/Rfc6238AuthenticationService.cs:15]: @MikeAlhayek I'm not sure if this the right direction to implement own `Rfc6238AuthenticationService` version. Original `Rfc6238AuthenticationService` service do not allow changing time span and pin length by design. **It is for security reason.**

Please imagine the scenario when somebody, will use 2 digits code and time span set to let's say to 10 minutes... You would be able to guess the code by trying multiple times all possible codes.

Other than that, I've extracted your service and adjusted it a bit (to be able to pass UtcNow) to test a few possible edge cases, and I found that something is wrong. I used 10 minutes time span, and after this time, I'm getting that my PIN is still valid:

your service tested:
image

original service tested:
image

I will try to find some time to read your code and check it deeper. But as I said, I'm not sure if this is the right way. We should use providers which allow using configurable timespans. Those providers are not good for all scenarios, because codes produced are much longer (so they are not suitable for SMS). It is because `creation time` is a part of encrypted (protected) code. You can read more [here](https://github.com/dotnet/aspnetcore/blob/79ef5e329b1e31c3775a1977798253cc8f7da6cc/src/Identity/Core/src/DataProtectorTokenProvider.cs#L116).

File: src/OrchardCore.Modules/OrchardCore.Users/Views/TemplateUserConfirmEmail.cshtml
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Users/Views/TemplateUserConfirmEmail.cshtml:8]: Is it something you do because there is a suggestion?

File: src/OrchardCore/OrchardCore.Users.Core/Services/Rfc6238AuthenticationService.cs
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Users.Core/Services/Rfc6238AuthenticationService.cs:14]: @lampersky I removed support for pins length < 6 characters. The Algorithm that generates pins validates with pics generated outside the expiration date-time because the clock that generates tokens could be different from the clock that validates the token. More about [RFC6238 ](https://datatracker.ietf.org/doc/html/rfc6238)

Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Users.Core/Services/Rfc6238AuthenticationService.cs]: Any way, I removed it and cleanup some other stuff.

File: src/OrchardCore.Modules/OrchardCore.Users/TwoFactorAuthenticationStartup.cs
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/TwoFactorAuthenticationStartup.cs:98]: @sebastienros the reason we create the type using `MakeGenericType` here is because we also need the type on line 102 when creating an instance of `TokenProviderDescriptor`.

Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.Users/TwoFactorAuthenticationStartup.cs:98]: so again `typeof(EmailTokenProvider<IUser>)` on line 102?

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/TwoFactorAuthenticationStartup.cs:98]: yes but why? Check this out https://github.com/dotnet/aspnetcore/blob/main/src/Identity/Core/src/IdentityBuilderExtensions.cs#L28-L38

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/TwoFactorAuthenticationStartup.cs:98]: @sebastienros do you still need me to update this?

**Pull request merged**
